Document ROS Iron discovery options

sloretz opened the pull request:
Documents new discovery options in ROS Iron

https://github.com/ros2/ros2/issues/1359

sloretz committed:
Document ROS Iron discovery configuration

Signed-off-by: Shane Loretz <sloretz@google.com>

fujitatomoya commented on source/Releases/Release-Iron-Irwini.rst:
```diff
@@ -81,6 +81,32 @@ There are examples of this in action in https://github.com/ros2/demos/blob/iron/
+Previous ROS 2 versions offered limited discovery options.
+The default behavior for DDS based RMW implementations was to discover any node reachable via multicast.
+It could be limited to the same machine by setting the environment variable `ROS_LOCALHOST_ONLY`, but any additional configuration required configuring the middleware directly, usually via middleware specific XML files and environment variables.
+ROS Iron retains the same default discovery behavior, but replaces `ROS_LOCALHOST_ONLY` with more granular options.
```
I think that `ROS_LOCALHOST_ONLY` is not fully replaced, just deprecated. so it is still honored if user sets it?

```suggestion
ROS Iron retains the same default discovery behavior, but deprecates `ROS_LOCALHOST_ONLY` with more granular options.
```

sloretz replied:
If I understand correctly, it's still being passed down to the RMW implementations, but `rmw_fastrtps*`, `rmw_cyclonedds_cpp`, and `rmw_connext` aren't doing anything with it. `rcl` [warns if it's set](./ros2/rmw_cyclonedds/rmw_cyclonedds_cpp/src/rmw_node.cpp). I suppose you're right it would be more correct to say it's deprecated since other middlewares could still choose to use it.

sloretz replied:
269a9f8d

sloretz committed:
replaces -> deprecates

Signed-off-by: Shane Loretz <sloretz@google.com>

clalancette commented on source/Releases/Release-Iron-Irwini.rst:
```diff
@@ -81,6 +81,32 @@ There are examples of this in action in https://github.com/ros2/demos/blob/iron/
+
+`ROS_AUTOMATIC_DISCOVERY_RANGE` controls how far ROS nodes will try to discover each other.
+Valid options are `SUBNET`, `LOCALHOST`, `OFF`, and `SYSTEM_DEFAULT`.
+`SUBNET` is the default, and for DDS based middlewares it means it will discovery any node reachable via multicast.
```
```suggestion
`SUBNET` is the default, and for DDS based middlewares it means it will discover any node reachable via multicast.
```

sloretz replied:
06219212

clalancette commented on source/Releases/Release-Iron-Irwini.rst:
```diff
@@ -81,6 +81,32 @@ There are examples of this in action in https://github.com/ros2/demos/blob/iron/
+Previous ROS 2 versions offered limited discovery options.
+The default behavior for DDS based RMW implementations was to discover any node reachable via multicast.
+It could be limited to the same machine by setting the environment variable `ROS_LOCALHOST_ONLY`, but any additional configuration required configuring the middleware directly, usually via middleware specific XML files and environment variables.
+ROS Iron retains the same default discovery behavior, but deprecates `ROS_LOCALHOST_ONLY` in favor of more granular options.
```
Hm, I didn't actually understand that `ROS_LOCALHOST_ONLY` is effectively being ignored now.  That means that it is not "deprecated", it is removed for all intents and purposes (though we still have a warning if a user sets it).

I'm wondering if we should restore honoring `ROS_LOCALHOST_ONLY` so we can tick-tock it, and then remove it in J-Turtle.  I think it would be relatively straightforward to do; we would just need to update the `rmw_*` implementations to also honor it if it was set.  Thoughts on this?

sloretz replied:
> I'm wondering if we should restore honoring ROS_LOCALHOST_ONLY so we can tick-tock it, and then remove it in J-Turtle. I think it would be relatively straightforward to do; we would just need to update the rmw_* implementations to also honor it if it was set. Thoughts on this?

I'm not sure the RMW implementations have enough info to respect `ROS_LOCALHOST_ONLY`. They only see the resulting [`localhost_only` and `discovery_options` in the `rmw_init_options_t`](https://github.com/ros2/rmw/blob/36003aaed8fd9718c20b19f4da6fb21929574058/rmw/include/rmw/init_options.h#L57-L59). Say range is set to `SUBNET` and `localhost_only` is `true`, what should the RMW implementation do? `localhost_only` was probably set by the user, but `SUBNET` could have been explicitly set by the user too.

The `rcl` layer might be able to do it, but it will need decisions about how to handle multiple sources. If we were only dealing with the environment variables, then it would be easy to say *"if `ROS_LOCALHOST_ONLY` is set and `ROS_AUTOMATIC_DISCOVERY_RANGE` is not, then use `ROS_LOCALHOST_ONLY` to set `ROS_AUTOMATIC_DISCOVERY_RANGE`"*. However, the user could have [set `rmw_init_options_t` programmatically before calling `rcl_init`](https://github.com/ros2/rcl/blob/a01d0951adbf346fe0acd4e4be24b896bc95bf50/rcl/include/rcl/init_options.h#L198-L199). If the `rmw_init_options` passed programatically says `localhost_only = true`, but environment variables for `ROS_AUTOMATIC_DISCOVERY_RANGE` and `ROS_STATIC_PEERS` are set, what should happen?

Maybe at the `rcl` layer we could say: *"if any source says `ROS_LOCALHOST_ONLY/localhost_only` is `true`, then unconditionally set automatic discovery range to `LOCALHOST`"*?

In the meantime, I'd recommend opening a ticket about it on `rcl`. We can update the release notes if someone changes the behavior before the release.

fujitatomoya replied:
I was expecting that if `localhost_only` is set, it prevails to `discovery_options` for Iron to keep the backward compatibility and giving some time for application adjustment. (that is why i suggested `deprecated` here)

but,

> ROS_LOCALHOST_ONLY is effectively being ignored now. That means that it is not "deprecated", it is removed for all intents and purposes (though we still have a warning if a user sets it).

this is current implementation, so `localhost_only` is never honored. it is no use even if it is set.

if we go with current implementation, `removed` is appropriate for the note? and probably we should update `rcl` warning message accordingly.
but i think this behavior change would be surprising if the application set `localhost_only`, it prints that `localhost_only is deprecated` warning but user would expect it still be effective.

how about when `localhost_only` is set, `automatic_discovery_range` falls back to `RMW_AUTOMATIC_DISCOVERY_RANGE_LOCALHOST` and `static_peers` to none? that can let the user know `localhost_only` is deprecated and will be removed in the future w/o surprise?

clalancette replied:
Yeah, I agree that we should make `localhost_only` essentially equivalent to `RMW_AUTOMATIC_DISCOVERY_RANGE_LOCALHOST`.  I know it isn't 100% the same, but at least we'll do something for it.

@fujitatomoya If you or anyone on your team has time to take a look at that, it would be most appreciated.  If not, I'll try to find some time to do it later on this week.

fujitatomoya replied:
@sloretz @clalancette 

https://github.com/ros2/rcl/pull/1069

fujitatomoya replied:
we need to update the description once https://github.com/ros2/rcl/pull/1069 is merged.

fujitatomoya replied:
I created tutorial documentation for this, https://github.com/ros2/ros2_documentation/pull/3596

sloretz committed:
discovery -> discover

Signed-off-by: Shane Loretz <sloretz@google.com>

fujitatomoya commented:
https://github.com/ros2/rcl/issues/1070 as follow-up to remove `localhost_only` for J-turtle.

clalancette committed:
Revamp of the discovery options.

Fix the highlighted areas, and also consolidate the language
down into bullet points.  I think this is easier to read.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
OK, with the work that @fujitatomoya did, this is now ready to get in.  I've rebased this and also revamped it to be bullet points to be easier to read.  I also fixed up all of the highlighted areas.

I'm going to approve this, since I think this is now ready to go.  But since I've done substantial work here, I'd like to get another approval before I merge.

fujitatomoya reviewed (approved):
lgtm
